gateway/shard/manager.go

```diff
@@ -0,0 +1,327 @@
+		NumShards:              numShards,
+		onShardingRequiredExec: new(moreatomic.Bool),
+	}
+
+	for _, g := range m.gateways {
+		g.Events = m.Events
+		g.OnScalingRequired = m.onGatewayScalingRequired
+	}
+
+	return m
+}
+
+// FromShardID returns the *gateway.Gateway with the given shard id, or nil if
+// the shard manager has no gateways with the given id.
+func (m *Manager) FromShardID(shardID int) *gateway.Gateway {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	// fast-path, also prevent nil pointer dereference if this manager manages
+	// a user account
+	if m.NumShards == 1 {
+		return m.gateways[0]
+	}
+
+	i := sort.Search(len(m.gateways), func(i int) bool {
+		return m.gateways[i].Identifier.Shard.ShardID() >= shardID
+	})
+
+	if i < len(m.gateways) && m.gateways[i].Identifier.Shard.ShardID() == shardID {
+		return m.gateways[i]
+	}
+
+	return nil
+}
+
+// FromGuildID returns the *gateway.Gateway managing the guild with the passed
+// ID, or nil if this Manager does not manage this guild.
+func (m *Manager) FromGuildID(guildID discord.GuildID) *gateway.Gateway {
+	return m.FromShardID(int(uint64(guildID>>22) % uint64(m.NumShards)))
+}
+
+// Apply applies the given function to all gateways handled by this Manager.
+func (m *Manager) Apply(f func(g *gateway.Gateway)) {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	for _, g := range m.gateways {
+		f(g)
+	}
+}
+
+// ApplyError is the same as Apply, but the iterator function returns an error.
+// If such an error occurs, the error will be returned wrapped in an *Error.
+//
+// If all is set to true, ApplyError will apply the passed function to all
+// gateways. If a single error occurs, it will be returned as an *Error, if
+// multiple errors occur then they will be returned as *MultiError.
+func (m *Manager) ApplyError(f func(g *gateway.Gateway) error, all bool) error {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	for _, g := range m.gateways {
+		if err := f(g); err != nil {
+			return &Error{
+				ShardID: shardID(g),
+				Source:  err,
+			}
+		}
+	}
+
+	return nil
+}
+
+// Gateways returns the gateways managed by this Manager.
+func (m *Manager) Gateways() []*gateway.Gateway {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	cp := make([]*gateway.Gateway, len(m.gateways))
+	copy(cp, m.gateways)
+
+	return cp
+}
+
+// Open opens all gateways handled by this Manager.
+// If an error occurs, Open will attempt to close all previously opened
+// gateways before returning.
+func (m *Manager) Open() error {
+	var errs MultiError
+
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	for i, g := range m.gateways {
+		if err := g.Open(); err != nil {
+			errs = append(errs, &Error{
+				ShardID: shardID(g),
+				Source:  err,
+			})
+
+			for _, g := range m.gateways[:i] {
+				if err := g.Close(); err != nil {
+					errs = append(errs, &Error{
+						ShardID: shardID(g),
+						Source:  err,
+					})
+				}
+			}
+
+			if len(errs) == 1 {
+				return errs[0]
+			}
+
+			return errs
+		}
+	}
```

Comment:
It actually can't, because unlike the other methods like `Close`, `Open` attempts to close all previously opened gateways if it encounters an error.
This was done, because in most cases if you `Open` your session/state during startup, and encounter an error, you would most likely exit the program.

Reply:
You can probably just loop over all gateways again to close them if `ApplyError` returned an error unless I'm missing something.

Reply:
No, because in an error scenario we would need the index of the gateway that returned an error, so we only close those gateways (i.e `m.gateways[:i]`, see line 248) that were previously opened.

Reply:
`ShardID` can provide that information. If the operation of searching based on ShardID is too expensive, it might be a sign that `Error` could use an "internal ID" to be more useful.

Reply:
Sure, I can do that. Computational expense is negligible since: a) open is ever only performed once, b) network is the true bottleneck here, and c) there will never be a large number of shards stored in a manager.